### PR TITLE
Add CMA-MAE Sampler (CmaMaeSampler)

### PR DESCRIPTION
**Pull request by btjanaka**
## Contributor Agreements

Please read the [contributor agreements](https://github.com/optuna/optunahub-registry/blob/main/CONTRIBUTING.md#contributor-agreements) and if you agree, please click the checkbox below.

- [x] I agree to the contributor agreements.

> [!TIP]
> Please follow the [Quick TODO list](https://github.com/optuna/optunahub-registry/tree/main?tab=readme-ov-file#quick-todo-list-towards-contribution) to smoothly merge your PR.

## Motivation

This PR creates the `CmaMaeSampler`, which provides the CMA-MAE algorithm as implemented in pyribs.

[CMA-MAE](https://arxiv.org/abs/2205.10752>) is a quality diversity algorithm that has demonstrated state-of-the-art performance in a variety of domains. [pyribs](https://pyribs.org>) is a bare-bones Python library for quality diversity optimization algorithms. For a primer on CMA-MAE and pyribs, we recommend referring to the series of [pyribs tutorials](https://docs.pyribs.org/en/stable/tutorials.html).

For simplicity, this implementation provides a default instantiation of CMA-MAE with a [GridArchive](https://docs.pyribs.org/en/stable/api/ribs.archives.GridArchive.html) and [EvolutionStrategyEmitter](https://docs.pyribs.org/en/stable/api/ribs.emitters.EvolutionStrategyEmitter.html) with improvement ranking, all wrapped up in a [Scheduler](https://docs.pyribs.org/en/stable/api/ribs.schedulers.Scheduler.html).

## Description of the changes

- Added `CmaMaeSampler` in `sampler.py`
- Added an example for using `CmaMaeSampler` in `example.py`

## TODO List towards PR Merge

Please remove this section if this PR is not an addition of a new package.
Otherwise, please check the following TODO list:

- [x] Copy `./template/` to create your package
- [x] Replace `<COPYRIGHT HOLDER>` in `LICENSE` of your package with your name
- [x] Fill out `README.md` in your package
- [x] Add import statements of your function or class names to be used in `__init__.py`
- [x] (Optional) Add `from __future__ import annotations` at the head of any Python files that include typing to support older Python versions
- [x] Apply the formatter based on the tips in [`README.md`](https://github.com/optuna/optunahub-registry/tree/main)
- [x] Check whether your module works as intended based on the tips in [`README.md`](https://github.com/optuna/optunahub-registry/tree/main)

### CODE REVIEW

**Comment by btjanaka** [comment]
Hi @nabenabe0928, I just started this PR for implementing the CMA-MAE algorithm in Optuna. I am wondering if I am on the right track. So far, I have set up the initialization function. I am wondering how I should handle `sample_relative` -- I believe there was a part we wanted to check due to the enqueuing of trials.

Currently, I am actually getting an error that I believe is related to `sample_relative`; roughly, it says
```
TypeError: argument of type 'FrozenTrial' is not iterable
```

I'm also available to meet and work through these errors together. Thanks!

**Comment by nabenabe0928** [comment]
@btjanaka 
Thank you for the PR!
I will first look into your problem and will get you back on how we should work on this problem!

**Comment by nabenabe0928** [comment]
@btjanaka 

I created an example code for batch optimization libraries!
Could you have a look and if you have any difficulties, please tell me!

```python
from __future__ import annotations

from typing import Iterable

import numpy as np
import optuna
import optunahub


_logger = optuna.logging.get_logger(f"optuna.{__name__}")
SimpleBaseSampler = optunahub.load_module("samplers/simple", force_reload=False).SimpleBaseSampler

class SamplerToWrap:
    def __init__(self, dim: int, batch_size: int, seed: int | None) -> None:
        self._batch_size = batch_size
        self._dim = dim
        self._rng = np.random.RandomState(seed)
        self._mean = np.zeros(dim, dtype=float)
        self._cov = np.identity(dim, dtype=float)

    def ask(self) -> np.ndarray:
        return self._rng.multivariate_normal(mean=self._mean, cov=self._cov, size=self._batch_size)

    def tell(self, params: np.ndarray, values: np.ndarray) -> None:
        assert len(values.shape) == 1
        assert len(params) == len(values)
        assert params.shape[-1] == self._dim
        # Take quantile so that at least two solutions will be considered as good.
        quantile = max(2 / len(values), 0.1)
        good_value = np.quantile(values, quantile)
        good_params = params[values <= good_value]
        # Take the statistics of good parameters.
        self._mean = np.mean(good_params, axis=0)
        self._cov = np.cov(good_params, rowvar=False)


class BatchOptSampler(SimpleBaseSampler):
    def __init__(self, dim: int, batch_size: int = 4, seed: int | None = None) -> None:
        _logger.warning("This sampler does not support parallel optimization now.")
        self._external_sampler = SamplerToWrap(batch_size=batch_size, dim=dim, seed=seed)
        self._batch_size = batch_size
        self._param_names = set([f"x{d}" for d in range(dim)])
        search_space = {
            name: optuna.distributions.FloatDistribution(-1e9, 1e9) for name in self._param_names
        }
        super().__init__(search_space=search_space)
        self._params_to_tell: list[np.ndarray] = []
        self._values_to_tell: list[np.ndarray] = []

    def _validate_param_names(self, given_param_names: Iterable[str]) -> None:
        if any(param_name not in self._param_names for param_name in given_param_names):
            raise ValueError("All param_name must be like ``x0``, ``x1``, ..., and so on.")

    def sample_relative(
        self, study: Study, trial: FrozenTrial, search_space: dict[str, BaseDistribution]
    ) -> dict[str, float]:
        self._validate_param_names(search_space.keys())
        params_array = self._external_sampler.ask()
        next_params = params_array[0]
        for params in params_array[1:]:
            study.enqueue_trial({name: params[d] for d, name in enumerate(self._param_names)})

        return {name: next_params[d] for d, name in enumerate(self._param_names)}

    def after_trial(
        self,
        study: Study,
        trial: FrozenTrial,
        state: TrialState,
        values: Sequence[float] | None,
    ) -> None:
        assert len(self._params_to_tell) == len(self._values_to_tell)
        assert len(values) == 1
        self._validate_param_names(trial.params.keys())
        self._params_to_tell.append([trial.params[name] for name in self._param_names])
        self._values_to_tell.append(values[0])
        if len(self._values_to_tell) != self._batch_size:
            return

        self._external_sampler.tell(
            params=np.asarray(self._params_to_tell), values=np.asarray(self._values_to_tell)
        )
        # Empty the results.
        self._params_to_tell = []
        self._values_to_tell = []


if __name__ == "__main__":
    def objective(trial: optuna.Trial) -> float:
        x0 = trial.suggest_float("x0", -5, 5)
        x1 = trial.suggest_float("x1", -5, 5)
        return x0**2 + x1**2

    sampler = BatchOptSampler(dim=2, batch_size=5)
    study = optuna.create_study(sampler=sampler)
    study.optimize(objective, n_trials=50)

```

File: package/samplers/cmamae/example.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+        emitter_batch_size=5,
+    )
+    study = optuna.create_study(
+        sampler=sampler,
+        directions=[
+            # pyribs maximizes objectives.
+            StudyDirection.MAXIMIZE,
+            # The remaining values are measures, which do not have an
+            # optimization direction.
+            # TODO: Currently, using StudyDirection.NOT_SET is not allowed as
```
Comment (btjanaka) [inline package/samplers/cmamae/example.py:40]: What direction should I use here? It seems Optuna only accommodates minimize and maximize. Quality diversity inherently involves a "no direction" optimization, hence the "diversity" aspect. Does it work to just use MINIMIZE as a placeholder?

File: package/samplers/cmamae/sampler.py [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+        return next_params
+
+    def after_trial(
+        self,
+        study: Study,
+        trial: FrozenTrial,
+        state: TrialState,
+        values: Sequence[float] | None,
+    ) -> None:
+        # TODO: Is it safe to assume the parameters will always come back in the
```
Comment (btjanaka) [inline package/samplers/cmamae/sampler.py:171]: Is it safe to assume the parameters will always come back in the order that they were sent out by the scheduler / sample_relative function? For example, if I send out solutions A, B, C, D, will they come back as A, B, C, D, or will they come back in some other order like D, B, C, A?

Pyribs makes the assumption that things are in order. If this is not the case, maybe we can retrieve solutions based on their trial ID?

File: package/samplers/cmamae/example.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+            # Optuna assumes we either minimize or maximize.
+            StudyDirection.MINIMIZE,
+            StudyDirection.MINIMIZE,
+        ],
+    )
+    study.optimize(objective, n_trials=100)
+
+    # TODO: Visualization.
```
Comment (btjanaka) [inline package/samplers/cmamae/example.py:48]: Are there any good examples showing integrations of other visualization tools? I recall we only discussed visualization briefly on our call.

**Comment by btjanaka** [comment]
@nabenabe0928 Thank you for sending the example code. I got it to work with pyribs, but I ran into a couple of small questions. I left them as comments on the PR; would you mind taking a look? Thank you!

File: package/samplers/cmamae/example.py (hunk 1 above)
Comment (nabenabe0928) [reply package/samplers/cmamae/example.py:37]: Usually, you would like to determine the sign of `values` for samplers and `directions` gives you the information for this.
If your sampler doesn't use this information, you can simply ignore this parameter, meaning that the default will be used.

Comment (btjanaka) [reply package/samplers/cmamae/example.py:37]: I see. The reason I looked into this is that the objective function for quality diversity algorithms like CMA-MAE needs to return multiple values, i.e., the objective and the measures. When I did not specify `directions`, I got an error that the objective function was only supposed to return one value.

**Comment by nabenabe0928** [comment]
Some updates will come after the PR below is merged:
https://github.com/btjanaka/optunahub-registry/pull/1

**Comment by btjanaka** [comment]
Hi @nabenabe0928, I think the PR is now ready for review! I will work on the visualization in a separate PR, but I did check that I could visualize the archive using the pyribs [grid_archive_heatmap](https://docs.pyribs.org/en/stable/api/ribs.visualize.grid_archive_heatmap.html) function and the results look good.

**Comment by nabenabe0928** [comment]
@btjanaka 
Hey, thank you for the update!
I will check your changes asap and will merge it!

**Review by nabenabe0928 — APPROVED** [review]
Thank you for your contribution, LGTM!
But for consistency, I will make another PR to follow the Optuna convention:)
If you, @btjanaka, want, please take a look at the PR!

**Pull request merged**